### DISCUSSION TITLE
A separate API to communicate BPP events

### DISCUSSION THREAD

**Discussion by pkapustin** [beckn/protocol-specifications#177] [Core Specification]
Events are a central part of the domain-driven design, event-driven architecture etc. They help a lot in building robust and scalable systems in a clean way.

For example, in KOMN taxi implementation, we have various events originating at the BPP side like ride assigned, ride started, ride ended, ride booking cancelled, etc. 

The events are essential in the implementations, and we think it would be best to represent them explicitly on the spec level. They allow to transmit just the necessary data associated with the event without any redundancy and ambiguity that would be present if one uses other types like Order, etc.

We suggest adding a new API, something like onEvent, and the specific event types can then be standardized on the network spec level.

**Comment by venkatramanm** [comment]
Great idea. 

We need to be clear though when on status , on cancel etc will be called if a status change or cancellation happened on BPP side? 

Or Are we saying we will migrate to this api for all events from BPP side.
 Or call all applicable APIs as we don't know what callbacks baps may implement?  

Generic api can have complex payloads. 
E.g item substitution, cancel, item cancel. Qty shortage etc.. we will need to think of the use cases and enrich the payload. Doable . Surely and I like the idea .
